### PR TITLE
Fixed - Default content &#8203; when adding gif and refactor gif module correctly

### LINKED ISSUE #1713: Default content &#8203; when adding gif and refactor gif module correctly
**Describe the bug**
When adding gif into activity, message or any other place where gif is supported, we add `&#8203;` into the content by default. Actually it is not needed because we already check for any validation when post is being added in ajax.

**To Reproduce**
1. Go to activity post form
2. Click on add gif and any add gif and update the post without any content.
3. Check the activity content in db.

**Expected behavior**
It should not automatically add `&#8203;` in content.

### PR DESCRIPTION
**Pull request by chetansatasiya**
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixes #1713 .

### How to test the changes in this Pull Request:

1. Go to activity post form
2. Click on add gif and any add gif and update the post without any content.
3. Check the activity content in db.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### API PR: 
https://github.com/buddyboss/buddyboss-platform-api/pull/265

### CODE REVIEW

File: src/bp-activity/bp-activity-functions.php [hunk 1]
```diff
@@ -2989,7 +2989,7 @@ function bp_activity_new_comment( $args = '' ) {
 	$feedback = __( 'There was an error posting your reply. Please try again.', 'buddyboss' );
 
 	// Bail if missing necessary data.
-	if ( empty( $r['content'] ) || empty( $r['user_id'] ) || empty( $r['activity_id'] ) ) {
+	if ( (empty( $r['content'] ) && empty($_POST['gif_data'])) || empty( $r['user_id'] ) || empty( $r['activity_id'] )  ) {
```
Comment (chetansatasiya) [inline src/bp-activity/bp-activity-functions.php:2992]: @jitendrabanjara1991 We need to remove this newly added condition.

File: src/bp-templates/bp-nouveau/includes/activity/ajax.php [hunk 2]
```diff
@@ -370,8 +370,8 @@ function bp_nouveau_ajax_new_activity_comment() {
 	if ( ! is_user_logged_in() ) {
 		wp_send_json_error( $response );
 	}
-
-	if ( empty( $_POST['content'] ) ) {
+       
+	if ( empty( $_POST['content'] ) && empty($_POST['gif_data']) ) {
```
Comment (chetansatasiya) [inline src/bp-templates/bp-nouveau/includes/activity/ajax.php:374]: @jitendrabanjara1991 We need to remove this newly added condition.

**Review by chetansatasiya — COMMENTED** [review]
@jitendrabanjara1991 Need to rework on this and we need to make whatever happening with the media same we have to implement here.

**Comment by jitendrabanjara1991** [comment]
@chetansatasiya 
When any new gif will added from comment and reply then it will store in DB with &nbsp. So, Please check it and let me know if anything else.
Here is the video - https://screencast-o-matic.com/watch/creXFcVcudI
Thanks.

File: src/bp-activity/bp-activity-functions.php [hunk 3]
```diff
@@ -2989,7 +2989,7 @@ function bp_activity_new_comment( $args = '' ) {
 	$feedback = __( 'There was an error posting your reply. Please try again.', 'buddyboss' );
 
 	// Bail if missing necessary data.
-	if ( empty( $r['content'] ) || empty( $r['user_id'] ) || empty( $r['activity_id'] ) ) {
+	if ( empty( $r['content'] ) || empty( $r['user_id'] ) || empty( $r['activity_id'] )  ) {
```
Comment (mehulkaklotar) [inline src/bp-activity/bp-activity-functions.php:2992]: @jitendrabanjara1991 extra space at the of the condition here.

File: src/bp-media/bp-media-filters.php [hunk 4]
```diff
@@ -66,7 +66,8 @@
 add_action( 'bp_activity_after_email_content', 'bp_media_activity_after_email_content' );
 
 add_filter( 'bp_get_activity_entry_css_class', 'bp_media_activity_entry_css_class' );
-
+//Check content empty or not for activity gif
```
Comment (mehulkaklotar) [inline src/bp-media/bp-media-filters.php:69]: please run your code changes though phpcs linting when you commit.

File: src/bp-templates/bp-nouveau/includes/activity/ajax.php [hunk 5]
```diff
@@ -370,8 +370,10 @@ function bp_nouveau_ajax_new_activity_comment() {
 	if ( ! is_user_logged_in() ) {
 		wp_send_json_error( $response );
 	}
+	
+	$content = apply_filters( 'bb_is_activity_content_empty', $_POST );
```
Comment (mehulkaklotar) [inline src/bp-templates/bp-nouveau/includes/activity/ajax.php:374]: I don't think we need to add &nbsp; here for activity if it contains gif only. we don't add &nbsp; for media and document when they added in activity if activity has no content in it. please remove this.

**Review by mehulkaklotar — CHANGES_REQUESTED** [review]
@jitendrabanjara1991 please do the changes requested.

**Comment by jitendrabanjara1991** [comment]
Hello @mehulkaklotar 
cc - @chetansatasiya 
As per your suggestion, We changed it. So, Please check it and let me know if anything else.
I also run phpcs only for my changes.
Here is the video - https://screencast-o-matic.com/watch/creoD8VcCRS
Thanks.

**Comment by jitendrabanjara1991** [comment]
Hello @mehulkaklotar
cc - @chetansatasiya
Fixed - When media, document, and gif add to the reply of the comment in the activity, then Space should be added instead of this content(#8203;).
Here is the video - https://screencast-o-matic.com/watch/creDbiVcJo3
Thanks.

**Comment by chetansatasiya** [comment]
@jitendrabanjara1991 Need to change as discussed introduce new filter and allow to create empty activity when media, document or gif added.

**Comment by jitendrabanjara1991** [comment]
@chetansatasiya 
cc - @KartikSuthar 
Fixed it. Here is the video - https://screencast-o-matic.com/watch/crhFICVfpq3
Please check it and let me know if anything else.
Thanks.

File: src/bp-media/bp-media-filters.php [hunk 6]
```diff
@@ -21,7 +21,7 @@
 add_action( 'bp_activity_comment_posted_notification_skipped', 'bp_media_activity_comments_update_media_meta', 10, 3 );
 add_action( 'bp_activity_after_delete', 'bp_media_delete_activity_media' ); // Delete activity medias.
 add_action( 'bp_activity_after_delete', 'bp_media_delete_activity_gif' ); // Delete activity gif.
-add_filter( 'bp_get_activity_content_body', 'bp_media_activity_embed_gif', 20, 2 );
+add_filter( 'bp_activity_entry_content', 'bp_media_activity_embed_gif' );
```
Comment (KartikSuthar) [inline src/bp-media/bp-media-filters.php:24]:
```suggestion
add_action( 'bp_activity_entry_content', 'bp_media_activity_embed_gif' );
```

File: src/bp-media/bp-media-filters.php [hunk 7]
```diff
@@ -1131,36 +1130,30 @@ function bp_media_activity_embed_gif_content( $activity_id ) {
...
-	$gif_content = bp_media_activity_embed_gif_content( $activity->id );
+	$gif_content = bp_media_activity_embed_gif_content( bp_get_activity_id() );
 
 	if ( ! empty( $gif_content ) ) {
-		$content .= $gif_content;
+		echo $gif_content;
 	}
-
-	return $content;
 }
```
Comment (KartikSuthar) [inline src/bp-media/bp-media-filters.php:1151]:
```
function bp_media_activity_embed_gif() {

	// check if profile and groups activity gif support enabled.
	if ( ( buddypress()->activity->id === bp_get_activity_object_name() && ! bp_is_profiles_gif_support_enabled() ) || ( bp_is_active( 'groups' ) && buddypress()->groups->id === bp_get_activity_object_name() && ! bp_is_groups_gif_support_enabled() ) ) {
		return false;
	}

	echo bp_media_activity_embed_gif_content( bp_get_activity_id() );
}
```

**Review by KartikSuthar — CHANGES_REQUESTED** [review]
@jitendrabanjara1991 Please do the following changes.

**Comment by jitendrabanjara1991** [comment]
@KartikSuthar 
cc - @chetansatasiya
Request changes done
Thanks.

**Review by KartikSuthar — APPROVED** [review]
@jitendrabanjara1991 Code looks good to me. Please merged to dev and pass to QA.

**Comment by jitendrabanjara1991** [comment]
@KartikSuthar 
cc - @mohit-buddyboss @sumitbuddyboss 
Merged in dev and ready for testing.
Thanks.

**Comment by mohit-buddyboss** [comment]
@jitendrabanjara1991 
. 
FIXED> Tested on dev branch of the platform.
Tested for the activity , activity comments , Group activity, group activity comments and for group forum and individual forum replies.
Tested this by uploading media, doc and GIF with or without text.
Now content (#(#8203;) is not appearing in database if user uploads only GIF instead of this blank entry appears. 

Activity and activity comments: https://www.loom.com/share/3dc9ec562bed4b239bf973cdb2d2b114
Group activity and comments : https://www.loom.com/share/e9f6b890e36240e8834fc5da1c241f8b
Edit activity and Forum replies : https://www.loom.com/share/2f097b0c28bc420d8e7e9c69c70419d2

In messages when we upload image, doc and GIF then it automatically add &nbsp; in content for that I created separate issue :https://github.com/buddyboss/buddyboss-platform/issues/2456

Related issue is resolve now and PR is ready for release.

cc @sourabhmatolia @chetansatasiya @KartikSuthar

**Comment by chetansatasiya** [comment]
@jitendrabanjara1991 fixed the conflict

**Comment by jitendrabanjara1991** [comment]
@chetansatasiya 
Fixed conflict.
Thanks

**Pull request merged**